README.md
# Python ElkM1 library

[![PyPI version](https://badge.fury.io/py/elkm1-lib.svg)](https://badge.fury.io/py/elkm1-lib)
[![CI](https://github.com/gwww/elkm1/actions/workflows/code-quality.yml/badge.svg)](https://github.com/gwww/elkm1/actions/workflows/code-quality.yml)
[![Downloads](https://pepy.tech/badge/elkm1-lib)](https://pepy.tech/project/elkm1-lib)
[![License](https://img.shields.io/github/license/mashape/apistatus.svg)](https://pypi.org/project/elkm1_lib/)

Library for interacting with ElkM1 alarm/automation panel.

https://github.com/gwww/elkm1

## Requirements

- Python 3.9 (or higher)

## Description

This package is created as a library to interact with an ElkM1 alarm/automation
pattern. The motivation to write this was to use with the Home Assistant
automation platform. The library can be used for writing other ElkM1 integration
applications. The IO with the panel is asynchronous over TCP or over the
serial port.

## Installation

```bash
    $ pip install elkm1_lib
```

## Overview

Basic connection to the Elk panel:

```python
    from elkm1_lib import Elk
    import logging

    # Print to STDOUT
    LOG = logging.getLogger(__name__)
    logging.basicConfig(level=logging.DEBUG, format='%(message)s')

    # Connect to elk
    elk = Elk({'url': 'elk://192.168.1.100'})
    elk.connect()
    elk.run()
```

The above will connect to the Elk panel at IP address 192.168.1.100. the `elk://`
prefix specifies that the connect is plaintext. Alternatively, `elks://` will
connect over TLS. In this case a userid and password must be specified
and the call to `Elk` changes to:

```python
    elk = Elk(
        {'url': 'elks://192.168.1.100', 'userid': 'test', 'password': 'pass'}
    )
```

The following ElkM1 connection protocols are supported:
- serial:// - Serial port;
- elk:// - Elk M1XEP Ethernet, non-secure
- elks:// - Elk M1XEP Ethernet, secure, TLS 1.0
- elksv1_0:// - Elk M1XEP Ethernet, secure, TLS 1.0, supported on M1XEP version < 2.0.46
- elksv1_2:// - Elk M1XEP Ethernet, secure, TLS 1.2, supported on M1XEP version = 2.0.46
- elksv1_3:// - Elk M1XEP Ethernet, secure, TLS 1.2, not yet supported on M1XEP, reserved for future

A username and password are required for any of the secure modes.

To see working example code take a look at the script `bin/simple`.

The `Elk` object supports the concept of `Elements`. An `Element`
is the base class representation of `Zones`, `Lights`, etc. So, for
example there is a list of zones: `elk.zones` and each zone can be
accessed by `elk.zones[index]`. Each element has a `__str__`
representation so that it is easy to print its contents.

All `Elements` are referenced starting at 0. Even though the Elk panel
refers to, for example, zones 1-208, the library references them
as zones 0-207. All translation from base 0 to 1 and vice-versa is
handled internally in the `elkm1_lib.message` module.

After creating the `Elk` object and connecting to the panel the
library code will synchronize all the elements to the data from the Elk panel.

Many Elk messages are handled by the library, caching their contents. When a
message causes a change to an attribute of an `Element`, registered
callbacks are called so that user use of the library can be notified
of changing elements. The following user code shows registering a callback:

```python
    def call_me(element, changeset):
       print(changeset)

    for zone_number in range(Max.ZONES.value):
      elk.zones[zone_number].add_callback(call_me)
```

The library encodes, decodes, and processes messages to/from the
Elk panel. All the encoding and decoding is done in `elkm1_lib.message` module.

Messages received are handled with callbacks. The library
internally registers callbacks so that decoded messages
can be used to update an `Element`. The user of the
library may also register callbacks. Any particular message
may have multiple callbacks.

When the message is received it is decoded
and some validation is done. The message handler is called
with the fields of from the decoded message. Each type of
message has parameters that match the message type. All handler parameters
are named parameters.

Here is an example of a message handler being registered and how it is called:

```python
    def zone_status_change_handler(zone_number, zone_status):
      print(zone_number, zone_status)

    elk.add_handler('ZC', zone_status_change_handler)
```

The above code registers a callback for 'ZC' (Elk zone status change)
messages. When a ZC message is received the handler functions are called
with the zone_number and zone_status.

There are a number of pseudo-handlers that act like the handlers. These are
called when events happen. The pseudo-handlers are:

- `connect`: When a successful connection to the ElkM1 is completed.
- `disconnect`: When a connection to a panel is disconnected.
- `login`: When a login is made to the panel (using `elks://` connection mode.
  A single boolean parameter is passed `succeeded`.
- `sync_complete`: When the panel has completed synchonizing all its elements.
- `timeout`: When a send of a message to the ElkM1 times out (fails to send).
- `unknown`: When a message from the ElkM1 is received and the library does
  not have a method to decode the message. The message is passed to this handler
  and can be decoded outside of the library.

## Utilities

The `bin` directory of the library has one utility program and
a couple of example uses of the library.

### `mkdoc`

The utility `mkdoc` creates a Markdown table of the list of Elk
messages with a check mark for those messages have encoders/decoders
and an X for those messages are not planned to be implemented.
There are no parameters to `mkdoc`. It outputs to stdout.
The data for the report comes from the ElkM1 library code mostly.
A couple of things are hard coded in the mkdoc script, notably
the "no plans to implement" list.

### `simple`

The `simple` Python script is a trivial use of the ElkM1 library.
It connects to the panel, syncs to internal memory, and continues
listening for any messages from the panel. The URL of the ElkM1 to
connect to is retrieved from an environment variable named `ELKM1_URL`.

### `elk`

The `elk` Python script is a bit of a command interpretor. It can run in
two modes. Non-interactive mode is the default. Just run the `elk` command.
The non-interactive mode is similar to `simple` except there are a
couple of message handlers (`timeout` and `unknown` handlers).

The `elk` can also be run in interactive mode by invoking it by
`elk -i`. In this mode is uses curses (full screen use of the terminal)
that has a command line and an output window. `TAB` switches between
the command line and output windows. In the output window the arrow keys
and scrollwheel scroll the contents of the window.

In the command line when running `elk -i` there are a
number of commands. Start with `help`. Then `help <command>` for
details on each command. In general there are commands to dump the internal
state of elements and to invoke any of the encoders to send a message
to the Elk panel.

For example, `light <4, 8, 12-14` would invoke the `__str__` method
for the light element to print the cached info for lights 0-3, 8, and 12-14.

Another example would be `pf 3` which issues the pf (Turn light off)
command for light number 3 (light 4 on the panel -- remember 0
versus 1 base).

All of the commands that send messages to the panel are automatically
discovered and are all the XX_encode functions in the ``elkm1_lib.message``
module. The docstring and the XX_encode's parameters are shown as part
of the help.

## Development

This project uses [poetry](https://poetry.eustace.io/) for development dependencies. Installation instructions are on their website.

To get started developing:

```
git clone https://github.com/gwww/elkm1.git
cd elkm1
poetry install
poetry shell # Or activate the created virtual environment
make test # to ensure everything installed properly
```

There is a `Makefile` in the root directory as well. The `make` command
followed by one of the targets in the `Makefile` can be used. If you don't
have or wish to use `make` the `Makefile` serves as examples of common
commands that can be run.

## Reporting a Bug

No problem ;) — report the bugs! But, logs are most often required. If you
are using Home Assistant, which is about the only use I'm aware of for
this library, then add the following to your `configuration.yaml`:
```
logger:
  default: info
  logs:
    custom_components.elkm1: debug
    elkm1_lib: debug
```

Do everything in your power to trim to logs down to their smallest. One way is
to reproduce your problem quickly so that few other logs are not generated in
between. Another recommendation is to use the simplest configuration that you
can think of to reproduce the problem.

Can you reproduce the problem in other ways? If this is a problem that is
being experienced while using Home Assistant try using the `Services` in `Developer Tools`.

Sometime logs may have sensitive information in them. You may want to
scan your logs for that info and "X" it out. In addition, you can send logs
directly to me. Support email is in the `pyproject.toml` file. You may also
send a link to somewhere you've stored/shared the file (DropBox for example).


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.
This project follows [Semantic Versioning](https://semver.org/).

## [2.2.1]
- Change minimum Python version back to 3.9

## [2.2.0]
- Refactor connection.py; no functional changes
- Refactor get_description; no functional changes
- Bump minimum Python version to 3.10

## [2.1.0]
- Add keypad function key and chime sync (Thx @jimboca)

## [2.0.2]
- Use a more liberal cipher set.

## [2.0.1]
- Support for Python 3.10; SSL cipher must be explicitly set

## [2.0.0]
- **Breaking change**: use Enums proper instead of Enum.values; signature of many
  of the xx_encode/xx_decode functions change as the parameters that were str,
  int, etc are now AlarmLevel, etc. Full list of of the enums used is in `const.py`
  Inspiration for this change from from Home Assistant move to using more
  enums to eliminate constants. Attribute changes:
  - Zone.definition (switch to Enum)
  - Zone.logical_status (switch to Enum)
  - Zone.physical_status (switch to Enum)
  - Area.armed_status (switch to Enum)
  - Area.arm_up_state (switch to Enum)
  - Area.alarm_status (switch to Enum)
  - Area.arm method uses ArmLevel instead of a string
  - Thermostat.mode (switch to Enum)
  - Thermostat.fan (switch to Enum)
  - Thermostat.set (use Enum)
  - Setting.value_format (switch to Enum)
  - Area.alarm_memory (changed from "0"/"1" string to bool)
- Add use of Generics on Elements to give better typing of self.elements

## [1.3.6]
- Refactor to separate concerns. Major part of this separation was pulling the
  observer code out of message.py and putting it in a separate class. This
  simplified message.py so that it is now just a set of encode/decode functions,
  no more class. This also made the connection.py code cleaner and focused
  on just the transport concerns.

## [1.3.5]
- Add logging of exception information on disconnect
- Add login event callback when first message received
- Move user/pass sending from connection to elk.py (cleaner connection code)

## [1.3.4]
- Fix wrong parameter introduced in typing exercise.

## [1.3.3]
- Remove pause/resume from `elk.py`; not used outside of internal lib and
  available in `Connection` if really needed
- Fix over zealous cleanup where password sent after sync. Doh.

## [1.3.2]
- Fix __iter__ type to be a Generator
- All the elements (Areas, Counters, Zones, etc) are now attributes of the
  Elk class instead of being dynamic attributes

## [1.3.1]
- Add py.typed
- Tweak CI to be more informative

## [1.3.0]
- Add typing: https://peps.python.org/pep-0561/
- Changed minimum Python version to 3.9
- Part of the typing change caused a circular import for username in util;
  Moved username to be a method on Users

## [1.2.2]
- On a non-secure connection still generate a login event on got_connection 
  (it will always be successful)

## [1.2.1]
- Bug in decode of UA message; did not handle hex for areas
- Cleanup after sync_complete to remove handler; allows for apps
  to send there own UA messages

## [1.0.0]
- Time to take this out of beta! The code has been stable for well over a year.
- Added support for TLS 1.2 secure connection mode. This is required for M1XEP
  version 2.0.46 and higher

## [0.8.11]
- Add get_voltage on zone.

## [0.8.8]
- Fix panel not recognized when using auto-configure.

## [0.8.7]
- Fix https://github.com/home-assistant/core/issues/20630

## [0.8.6]
- Many small non-functional code cleanups.

## [0.8.5]
- Lint and isort cleanup
- Move from .format to f-string for string formatting

## [0.8.4]
- Black formatting
- configured flag for each element set based on whether element has
  a description

## [0.8.3]
- Added rw (set time) support and RR (realtime clock) support
- Added callbacks when get connected and disconnected from panel
- KC now supports keycode 0

## [0.8.2]
- Added zone bypass helper and area bypass helper
- Zone bypass handler bug fix (did not handle bypass all)
- Renamed zone_trigger to trigger (breaking change)
- Fix comment typo

## [0.8.1]
- Duh, left breakpoint in code; removed
- Update dependencies (pytz left in poetry.lock)

## [0.8.0]
- Beefed up README with error reporting.
- Improve LD handling by simplifying and adding feature to note last_user

## [0.7.15]
- Switch to `poetry` from `pipenv` for dependency management; allowed
  removal of __version.py__, setup.py, setup.cfg, and of course
  Pipfile*

## [0.7.14]
- Support multiple elk instances in same process by associating
  _message_handlers, _sync_handlers, and the description processing
  code with elk object rather than as file globals.
- Moved message encoders/decoders into new MessageDecode class
  to facilitate the above.

## [0.7.13]
- Fix updating counter value
- Add alarm memory processing

## [0.7.12]
- Add system_trouble_status decoding.

## [0.7.11]
- Add pytz as dependency
- Add KC (keypress) handling

## [0.7.10]
- Add level() helper to control lights
- Remove turn_on and turn_off for lights

## [0.7.9]
- Add time of change to IC message. The will force callbacks to be called.
  Useful when multiple IC messages with same content come in. This way the
  app using the callback can record the multiple attempts.
- Add username utility function to get a user's name from an user number

## [0.7.8]
- Add zt decode and helper
- Add dm helper
- Fix speak word. Was using wrong Elk message code.

## [0.7.7]
- Add 30 second timeout to create connection
- lint cleanups

## [0.7.6]
- Set user name to default name if not configured
- Enhance IC decode to handle prox
- IC handler now saves user code
- Heartbeat bug; connect was called on heartbeat timeout AND then again on
  disconnect callback
- Made log messages on connect/disconnect/error clearer and more consistent
- Change max time on connection retry to 60 seconds (was 120 seconds)
- Tidied up a couple of comments

## [0.7.5]
- Add heartbeat

## [0.7.4]
- Fix triggered_alarm attr to zone

elkm1/elk.py
"""Master class that combines all ElkM1 pieces together."""

import asyncio
import logging
from importlib import import_module
import serial_asyncio

from .message import message_decode
from .proto import Connection
from .util import call_sync_handlers, parse_url, url_scheme_is_secure


LOG = logging.getLogger(__name__)

class Elk:
    """Represents all the components on an Elk panel."""

    def __init__(self, config, loop=None):
        """Initialize a new Elk instance."""
        self.loop = loop if loop else asyncio.get_event_loop()
        self._config = config
        self._conn = None
        self.connection_lost_callbk = None
        self._connection_retry_timer = 1

        # Setup for all the types of elements tracked
        if 'element_list' in config:
            self.element_list = config['element_list']
        else:
            self.element_list = ['panel', 'zones', 'lights', 'areas',
                                 'tasks', 'keypads', 'outputs', 'thermostats',
                                 'counters', 'settings', 'users']
        for element in self.element_list:
            self._create_element(element)

    def _create_element(self, element):
        module = import_module('elkm1.'+element)
        class_ = getattr(module, element.capitalize())
        setattr(self, element, class_(self))

    async def _connect(self, connection_lost_callbk=None):
        """Asyncio connection to Elk."""
        self.connection_lost_callbk = connection_lost_callbk
        url = self._config['url']
        LOG.debug("Elk connect to %s", url)
        scheme, dest, param, ssl_context = parse_url(url)
        try:
            if scheme == 'serial':
                # pylint: disable=C0330
                _coro = await serial_asyncio.create_serial_connection(self.loop,
                            lambda: Connection(self.loop,
                            self._connected, self._disconnected, self._got_data),
                            dest, baudrate=param)
            else:
                # pylint: disable=C0330
                _coro = await self.loop.create_connection(
                        lambda: Connection(self.loop,
                        self._connected, self._disconnected, self._got_data),
                        host=dest, port=param, ssl=ssl_context)
        except:
            LOG.debug("Connection failed. Retrying in %d seconds",
                      self._connection_retry_timer)
            self.loop.call_later(self._connection_retry_timer, self.connect)
            self._connection_retry_timer = 2 * self._connection_retry_timer \
                if self._connection_retry_timer < 64 else 120

    def _connected(self, _transport, conn):
        """Login and sync the ElkM1 panel to memory."""
        self._conn = conn
        self._connection_retry_timer = 1
        if url_scheme_is_secure(self._config['url']):
            self._conn.write_data(self._config['userid'], raw=True)
            self._conn.write_data(self._config['password'], raw=True)
        #self.loop.call_later(30, call_sync_handlers)
        call_sync_handlers()

    def _disconnected(self):
        LOG.debug("elk disconnected callback")
        self._conn = None
        self.loop.call_later(self._connection_retry_timer, self.connect)

    def _got_data(self, data): # pylint: disable=no-self-use
        LOG.debug("got_data '%s'", data)
        try:
            message_decode(data)
        except ValueError as err:
            LOG.debug(err)

    def is_connected(self):
        """Status of connection to Elk."""
        return self._conn is not None

    def connect(self):
        """Connect to the panel"""
        coro = self._connect()
        asyncio.ensure_future(coro)

    def run(self):
        """Enter the asyncio loop."""
        self.loop.run_forever()

    def send(self, msg):
        """Send a message to Elk panel."""
        self._conn.write_data(msg.message, msg.response_command)

    def pause(self):
        """Pause the connection from sending/receiving."""
        self._conn.pause()

    def resume(self):
        """Restart the connection from sending/receiving."""
        self._conn.resume()
